src/rgw/driver/sfs/multipart.cc

```diff
@@ -13,301 +13,616 @@
 
-  auto parts_it = parts.cbegin();
-  auto etags_it = part_etags.cbegin();
+    } else if (!part.etag.has_value()) {
+      lsfs_dout(
+          dpp, -1
+      ) << fmt::format("BUG: Part {} is finished and should have an etag!", k)
+        << dendl;
+      return -ERR_INTERNAL_ERROR;
+    }
 
-  for (; parts_it != parts.cend() && etags_it != part_etags.cend();
-       ++parts_it, ++etags_it) {
-    ceph_assert(etags_it->first >= 0);
-    if (parts_it->first != (uint32_t)etags_it->first) {
-      // mismatch part num
-      lsfs_dout(dpp, 0) << "mismatch part num, expected: " << parts_it->first
-                        << ", got " << etags_it->first << dendl;
+    ceph_assert(part.etag.has_value());
+    auto part_etag = part.etag.value();
+    auto etag = rgw_string_unquote(v);
+    if (part_etag != etag) {
+      lsfs_dout(dpp, 1)
+          << fmt::format(
+                 "client-specified part {} etag mismatch; expected {}, got {}",
+                 k, part_etag, etag
+             )
+          << dendl;
       return -ERR_INVALID_PART;
     }
 
-    auto part = parts_it->second;
-    auto part_it_etag = rgw_string_unquote(etags_it->second);
-    if (part->etag != part_it_etag) {
-      lsfs_dout(dpp, 0) << "mismatch part etag, expected: " << part->etag
-                        << ", got " << part_it_etag << dendl;
+    if ((part.len < 5 * 1024 * 1024) &&
+        (std::distance(it, part_etags.cend()) > 1)) {
+      lsfs_dout(dpp, 1) << fmt::format(
+                               "part {} is too small and not the last part!", k
+                           )
+                        << dendl;
+      return -ERR_TOO_SMALL;
+    }
+
+    to_complete[k] = p->second;
+    hash.update(part_etag);
+    expected_size += part.len;
+  }
+
+  if (store->filesystem_stats_avail_bytes.load() < expected_size) {
+    lsfs_dout(dpp, -1) << fmt::format(
+                              "filesystem stat reservation check hit. "
+                              "avail_bytes: {}, avail_pct: "
+                              "{}, total_bytes: {}, expected size: {}",
+                              store->filesystem_stats_avail_bytes,
+                              store->filesystem_stats_avail_percent,
+                              store->filesystem_stats_total_bytes, expected_size
+                          )
+                       << dendl;
+    return -ERR_QUOTA_EXCEEDED;
+  }
+
+  // calculate final etag
+  std::string etag = fmt::format("{}-{}", hash.final(), part_etags.size());
+
+  lsfs_dout(dpp, 10) << fmt::format(
+                            "upload_id: {}, final etag: {}", upload_id, etag
+                        )
+                     << dendl;
+
+  // start aggregating final object
+  res = mpdb.mark_aggregating(upload_id);
+  ceph_assert(res == true);
+
+  std::filesystem::path objpath =
+      store->get_data_path() / UUIDPath(mp->obj_uuid).to_path();
+
+  int objfd = ::open(objpath.c_str(), O_WRONLY | O_BINARY | O_CREAT, 0600);
```

Comment:
all that _should_ be true, but isn't at the moment. Once we move to that state, there's no way to do a complete again, because the path has marked the upload as complete.

That does make sense. I will try to find a way to do this in a simple way, but I might instead ask to push it to a later path set. This feels like a large chunk of changes.